### DISCUSSION TITLE
Joined Relations with $queryRaw

### DISCUSSION THREAD

**Discussion by bellybuttonbackwards** [prisma/prisma#17158] [Q&A]
Hello,

I have a hierarchical data structure, you can think of it as Folders with Files, using two tables. A Folder table has a self-referential relationship to capture the hierarchy and a File table has an FK relationship to a specific Folder.

eg
FOLDER
folderId: 1, parent: null, name: Root
folderId: 2, parent: 1, name: Child

FILE
fileId: 3, folderId: 2, name: File1
fileId: 4, folderId: 2, name: File2

I've written a `$queryRaw` using Postgres's `with recursive` that will return all folders under a given parent, as suggested by various support tickets, which works great. Is it possible to also pull back the Files and have them correctly modeled with Files as an array member of a single Folder using that same `$queryRaw` with a join, or do i need to do post processing on the json rows to collapse things correctly?

What I get:
```
[ { folderId: 1, parent: null, name: Root }, 
{ folderId: 2, parent: 1, name: Child, fileId: 3, folderId: 2, name: File1 },
{ folderId: 2, parent: 1, name: Child, fileId: 4, folderId: 2, name: File2 }
]
```
What I want:
```
[ { folderId: 1, parent: null, name: Root, files: [] }
{ { folderId: 2, parent: 1, name: Child, files: [ 
    { fileId: 3, folderId: 2, name: File1},
    { fileId: 4, folderId: 2, name: File2}]
 }]
```

https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#typing-queryraw-results makes reference to correct typing of relations 
> If you are selecting specific fields of the model or want to include relations, refer to the documentation about [leveraging Prisma Client's generated types](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety) if you want to make sure that the results are properly typed.

but it's unclear to me what scope of returned entity that applies to. I'd love to hear it's just a matter of naming the columns correctly so Prisma can intuit what I'm trying to do here. Is this possible?

Thanks,

matt

**Comment by pantharshit00** [comment]
Hello @scumbkt19,

Can you please share the query that you are using and the datamodel if possible? That way we can try to replicate your setup locally and try to resolve this.

**Reply by bellybuttonbackwards** [reply]
Thank you for the response @pantharshit00 , i think in trying to explain why i needed to make a $rawQuery I distracted from the core of my question. To simplify, let's say I have a simple 1-1 parent child relationship

```
model Folder {
  id String  @db.Uuid
  name String
}

model File {
  id String @db.Uuid
  name String
  folderId String @db.Uuid
  folder   Folder @relation(fields: [folderId], references: [id])
}
```

for reasons, I need to query using a $rawQuery, but I'd like to pull back both the File and its linked Folder in a single query, something like

```
select file.*, folder.*
from File file
left join Folder folder
on file.folderId = folder.id
where file.id = 1
```
which generates one wide response row with every column from File then every column from Folder


If i was able to use the standard `findUnique` that would be typed as a `File & { folder: Folder}`

The documentation I linked says that I can use the parameter typed $rawQuery the same way https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#typing-queryraw-results

specifically:
> If you are selecting specific fields of the model or want to include relations

I want to include a relation, and I want it to be typed, but I haven't figured out how to actually get Prisma to model relations on the $rawQuery. How do I do that?

Thanks,

matt

**Reply by ludralph** [reply]
Hi @scumbkt19 

You can type the results of a `$queryRaw` to match the expected return type of a Prisma model, Prisma does not automatically handle the nesting of related records when using raw queries. The `$queryRaw` method will return the data as it is retrieved from the database, which means that if you perform a join in your raw SQL query, the result will be a flat structure with columns from both tables.

You can specify a type when invoking $queryRaw to ensure type safety. However, this does not imply that Prisma will automatically nest the related records for you.

`PrismaPromise<T>` uses a generic type parameter `T`. You can determine the type of `T` when you invoke the `$queryRaw` method. In the following example, `$queryRaw` returns `User[]`:

// import the generated `User` type from the `@prisma/client` module
```js
import { User } from '@prisma/client'
const result = await prisma.$queryRaw<User[]>`SELECT * FROM User`
// result is of type: `User[]`
```
To achieve the nested structure you desire, you would need to manually map the flat results to a nested JSON structure in your application code after fetching the results with `$queryRaw`. This might involve creating a new object for each file and attaching the related folder information to it.
If you want to ensure that the results are properly typed, you could define a custom type that represents the structure of the joined data and use that type as the generic parameter for $queryRaw. However, you would still need to perform the manual mapping to achieve the nested structure.
For example:
```js
type FileWithFolder = {
  fileId: string;
  fileName: string;
  folderId: string;
  folderName: string;
  // ... other fields
};
const result = await prisma.$queryRaw<FileWithFolder[]>`SELECT file.*, folder.* FROM File file LEFT JOIN Folder folder ON file.folderId = folder.id WHERE file.id = 1`;
// result is of type: `FileWithFolder[]`
// Manual mapping to nest folder data inside file data would be required here

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
